[Navigation] Fix unstable UUIDs in failing WPT tests

rwlbuis opened the pull request:
#### f382ec34d071dcbf3623d1d2fc252b5806c7aa04
<pre>
[Navigation] Fix unstable UUIDs in failing WPT tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=278253">https://bugs.webkit.org/show_bug.cgi?id=278253</a>

Reviewed by Alex Christensen and Darin Adler.

In order to fix the unstable UUIDs in failing WPT tests various improvements of handling NavigationHistoryEntry.key are done
- same document navigation history items keep their navigation API key
- cross document navigations that are same origin keep their navigation API keys (1)
- we rely on the navigation API at the time the NavigationHistoryEntry was created and do not fall back to the associated
  history item.

[1] <a href="https://html.spec.whatwg.org/multipage/browsing-the-web.html#finalize-a-cross-document-navigation">https://html.spec.whatwg.org/multipage/browsing-the-web.html#finalize-a-cross-document-navigation</a> (Step 9.3)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-srcdoc-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-replace-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-replace-same-document-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::reset):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::setUUIDIdentifier):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadInSameDocument):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::updateCurrentItem):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::create):
(WebCore::NavigationHistoryEntry::key const):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: <a href="https://commits.webkit.org/285078@main">https://commits.webkit.org/285078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/045d4bdb82439045b7e7373fbf0136ee73c3929a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71477 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24250 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75588 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22682 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73592 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58690 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56448 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14921 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74543 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46183 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61574 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36897 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42846 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19038 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21023 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64747 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19404 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77307 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15711 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18577 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/64168 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15754 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61613 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64164 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15804 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12311 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5948 "Build is in progress. Recent messages:OS: Ventura (13.6.9), Xcode: 14.3; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed layout tests; 9 flakes") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46690 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1469 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47761 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49045 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47503 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -888,6 +888,8 @@ void HistoryController::updateCurrentItem()
         // dependent on the document.
         bool isTargetItem = currentItem->isTargetItem();
+        if (!isEqualIgnoringQueryAndFragments(currentItem->url(), documentLoader->url()))
+            currentItem->regenerateUUIDIdentifier();
```
I'd put this in `HistoryItem::reset()`

rwlbuis replied:
Thanks, but I now think this can be solved differently/better in another way. I'll keep the PR open for now as a reminder.

darinadler commented on Source/WebCore/page/NavigationHistoryEntry.cpp:
```diff
@@ -40,7 +40,7 @@ namespace WebCore {
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(NavigationHistoryEntry);
 
-NavigationHistoryEntry::NavigationHistoryEntry(ScriptExecutionContext* context, Ref<HistoryItem>&& historyItem, String urlString, RefPtr<SerializedScriptValue>&& state, WTF::UUID key, WTF::UUID id)
+NavigationHistoryEntry::NavigationHistoryEntry(ScriptExecutionContext* context, Ref<HistoryItem>&& historyItem, String urlString,  WTF::UUID key, RefPtr<SerializedScriptValue>&& state, WTF::UUID id)
```
There’s a stray space here before `WTF::`.

rwlbuis replied:
Well spotted, fixed.

darinadler commented on Source/WebCore/page/NavigationHistoryEntry.h:
```diff
@@ -60,7 +60,7 @@ class NavigationHistoryEntry final : public RefCounted<NavigationHistoryEntry>,
 
 private:
-    NavigationHistoryEntry(ScriptExecutionContext*, Ref<HistoryItem>&&, String urlString, RefPtr<SerializedScriptValue>&& state = { }, WTF::UUID key = WTF::UUID::createVersion4(), WTF::UUID = WTF::UUID::createVersion4());
+    NavigationHistoryEntry(ScriptExecutionContext*, Ref<HistoryItem>&&, String urlString,  WTF::UUID key, RefPtr<SerializedScriptValue>&& state = { }, WTF::UUID = WTF::UUID::createVersion4());
```
There’s a stray space here before `WTF::`.

rwlbuis replied:
Done.

rwlbuis committed:
[Navigation] Fix unstable UUIDs in failing WPT tests
https://bugs.webkit.org/show_bug.cgi?id=278253

Reviewed by Alex Christensen and Darin Adler.

In order to fix the unstable UUIDs in failing WPT tests various improvements of handling NavigationHistoryEntry.key are done
- same document navigation history items keep their navigation API key
- cross document navigations that are same origin keep their navigation API keys (1)
- we rely on the navigation API at the time the NavigationHistoryEntry was created and do not fall back to the associated
  history item.

[1] https://html.spec.whatwg.org/multipage/browsing-the-web.html#finalize-a-cross-document-navigation (Step 9.3)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/entries-after-srcdoc-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/key-id-location-replace-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-replace-same-document-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/history/HistoryItem.cpp:
(WebCore::HistoryItem::reset):
* Source/WebCore/history/HistoryItem.h:
(WebCore::HistoryItem::setUUIDIdentifier):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadInSameDocument):
* Source/WebCore/loader/HistoryController.cpp:
(WebCore::HistoryController::updateCurrentItem):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeForNewWindow):
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::create):
(WebCore::NavigationHistoryEntry::key const):
* Source/WebCore/page/NavigationHistoryEntry.h:

Canonical link: https://commits.webkit.org/285078@main
